tree/dataframe/src/RNTupleDS.cxx

```diff
@@ -69,27 +133,120 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
+   // "__rdf_sizeof_event.tracks.hits.x"  [std::vector<unsigned int>]
+   // "event.tracks.hits.y"               [std::vector<std::vector<float>>]
+   // "__rdf_sizeof_event.tracks.hits.y"  [std::vector<unsigned int>]
+
+   const auto &fieldDesc = desc.GetFieldDescriptor(fieldId);
+   if (fieldDesc.GetStructure() == ENTupleStructure::kCollection) {
+      // Inner fields of collections are provided as projected collections of only that inner field,
+      // E.g. we provide a projected collection vector<vector<float>> for "event.tracks.hits.x" in the example
+      // above.
+
+      // We open a new collection scope with fieldID being the inner most collection. E.g. for "event.tracks.hits",
+      // skeinIDs would already contain the fieldID of "event.tracks"
+      skeinIDs.emplace_back(fieldId);
+      // There should only be one sub field but it's easiest to access via the sub field range
+      for (const auto& f : desc.GetFieldRange(fieldDesc.GetId())) {
+         AddField(desc, colName, f.GetId(), skeinIDs);
+      }
+      // Note that at the end of the recursion, we handled the inner sub collections as well as the
+      // collection as whole, so we are done.
+      return;
+   } else if (fieldDesc.GetStructure() == ENTupleStructure::kRecord) {
+      // Inner fields of records are provided as individual RDF columns, e.g. "event.id"
+      for (const auto& f : desc.GetFieldRange(fieldDesc.GetId())) {
+         auto innerName = colName.empty() ? f.GetFieldName() : (std::string(colName) + "." + f.GetFieldName());
+         AddField(desc, innerName, f.GetId(), skeinIDs);
+      }
    }
+
+   // The fieldID could be the root field or the class of fieldId might not be loaded.
+   // In these cases, only the inner fields are exposed as RDF columns.
+   auto fieldOrException = Detail::RFieldBase::Create("", fieldDesc.GetTypeName());
+   if (!fieldOrException)
+      return;
+   auto valueField = fieldOrException.Unwrap();
+   valueField->SetOnDiskId(fieldId);
+   std::unique_ptr<Detail::RFieldBase> cardinalityField;
+   // Collections get the additional "number of" RDF column (e.g. "__rdf_sizeof_tracks")
+   if (!skeinIDs.empty()) {
+      cardinalityField = std::make_unique<Detail::RRDFCardinalityField>();
+      cardinalityField->SetOnDiskId(skeinIDs.back());
+   }
+
+   std::string typeName;
+   for (auto i = skeinIDs.rbegin(); i != skeinIDs.rend(); ++i) {
+      valueField = std::make_unique<ROOT::Experimental::RVectorField>("", std::move(valueField));
+      valueField->SetOnDiskId(*i);
+      // Skip the inner-most collection level to construct the cardinality column
+      if (i != skeinIDs.rbegin()) {
+         cardinalityField = std::make_unique<ROOT::Experimental::RVectorField>("", std::move(cardinalityField));
+         cardinalityField->SetOnDiskId(*i);
+      }
+   }
+
+   if (cardinalityField) {
+      fColumnNames.emplace_back(std::string("__rdf_sizeof_") + std::string(colName));
```

Comment:
Are you sure this works? I often see issue adding an `std::string` to a C string literal because the compiler thinks I want to add an `std::string` to a pointer

Reply:
yep `operator+` between std::string and string literal is "commutative"